File: src/org/softeg/slartus/forpdaplus/listfragments/TopicUtils.java
package org.softeg.slartus.forpdaplus.listfragments;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.widget.Toast;

import org.softeg.slartus.forpdaapi.Topic;
import org.softeg.slartus.forpdaapi.TopicApi;
import org.softeg.slartus.forpdaplus.Client;
import org.softeg.slartus.forpdaplus.MyApp;
import org.softeg.slartus.forpdaplus.R;
import org.softeg.slartus.forpdaplus.classes.AlertDialogBuilder;
import org.softeg.slartus.forpdaplus.classes.ThemeOpenParams;
import org.softeg.slartus.forpdaplus.classes.TopicListItemTask;
import org.softeg.slartus.forpdaplus.classes.forum.ExtTopic;
import org.softeg.slartus.forpdaplus.common.Log;
import org.softeg.slartus.forpdaplus.db.TopicsHistoryTable;

import java.io.IOException;

/**
 * Created by slinkin on 20.02.14.
 */
public class TopicUtils {
    /**
     * С какими параметрами навигации юзер решил открывать топик:  view=getlastpost и тд
     */
    public static String getOpenTopicArgs(CharSequence topicId,CharSequence template) {
        String themeActionPref = getTopicNavigateAction(template);
        return getUrlArgs(topicId,themeActionPref, null);
    }

    /**
     * Какой тип навигации юзер выбрал:  getlastpost, getfirstpost и тд
     */
    public static String getTopicNavigateAction(CharSequence template) {
        return PreferenceManager.getDefaultSharedPreferences(MyApp.getContext()).getString(String.format("%s.navigate_action", template), null);
    }

    public static void saveTopicNavigateAction(CharSequence template, CharSequence navigateAction) {
        PreferenceManager.getDefaultSharedPreferences(MyApp.getContext())
                .edit()
                .putString(String.format("%s.navigate_action", template), navigateAction.toString())
                .commit();
    }

    public static final String BROWSER = "browser";

    public static String getUrlArgs(CharSequence topicId, String openParam, String defaultUrlParam) {
        if (openParam == null) return defaultUrlParam;
        if (openParam.equals(ThemeOpenParams.BROWSER))
            return "";
        if (openParam.equals(Topic.NAVIGATE_VIEW_FIRST_POST))
            return "";
        if (openParam.equals(Topic.NAVIGATE_VIEW_LAST_POST))
            return "view=getlastpost";
        if (openParam.equals(Topic.NAVIGATE_VIEW_NEW_POST))
            return "view=getnewpost";
        if (openParam.equals(Topic.NAVIGATE_VIEW_LAST_URL)){
            try {
                String historyTopicUrl= TopicsHistoryTable.getTopicHistoryUrl(topicId);
                return TextUtils.isEmpty(historyTopicUrl)?
                        "view=getlastpost"
                        :
                        Uri.parse(historyTopicUrl).getQuery().replaceAll("showtopic=\\d+&?","");
            } catch (IOException e) {
                e.printStackTrace();
                return "view=getlastpost";
            }
        }


        return defaultUrlParam;

    }

    public static void showNavigateDialog(final Activity activity, final CharSequence templateId,
                                          final CharSequence topicId,
                                          final DialogInterface.OnClickListener onClickListener) {
        if (activity == null)
            return;

        CharSequence[] titles = new CharSequence[]{activity.getString(R.string.navigate_getfirstpost),
                activity.getString(R.string.navigate_getlastpost), activity.getString(R.string.navigate_getnewpost)};
        final CharSequence[] values = new CharSequence[]{Topic.NAVIGATE_VIEW_FIRST_POST,
                Topic.NAVIGATE_VIEW_LAST_POST, Topic.NAVIGATE_VIEW_NEW_POST};
        final int[] selected = {2};
        new AlertDialogBuilder(activity)
                .setSingleChoiceItems(titles, selected[0], new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        selected[0] = i;
                    }
                })
                .setTitle("Действие по умолчанию")
                .setPositiveButton("Всегда",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                dialog.dismiss();

                                new AlertDialogBuilder(activity)
                                        .setTitle("Подсказка")
                                        .setMessage("Вы можете изменить действие по умолчанию долгим тапом по теме")
                                        .setCancelable(false)
                                        .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialogInterface, int i) {
                                                dialogInterface.dismiss();

                                                String navigateAction = values[selected[0]].toString();
                                                TopicUtils.saveTopicNavigateAction(templateId, navigateAction);
                                                ExtTopic.showActivity(activity, topicId,
                                                        ThemeOpenParams.getUrlParams(navigateAction, null));

                                                onClickListener.onClick(null, -1);
                                            }
                                        })
                                        .create().show();
                            }
                        }
                )
                .setNeutralButton("Только сейчас",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                dialog.dismiss();

                                String navigateAction = values[selected[0]].toString();
                                ExtTopic.showActivity(activity, topicId,
                                        ThemeOpenParams.getUrlParams(navigateAction, null));

                                onClickListener.onClick(null, -1);
                            }
                        }
                )
                .create()
                .show();

    }

    public static String getTopicUrl(String topicId, String urlArgs) {
        return "http://4pda.ru/forum/index.php?showtopic=" + topicId + (TextUtils.isEmpty(urlArgs) ? "" : ("&" + urlArgs));
    }

    public static void showSubscribeSelectTypeDialog(final Context context, final android.os.Handler handler,
                                                     final CharSequence topicId) {
        showSubscribeSelectTypeDialog(context, handler, topicId,null);
    }
    public static void showSubscribeSelectTypeDialog(final Context context, final android.os.Handler handler,
                                                     final CharSequence topicId, final TopicListItemTask topicListItemTask) {
        CharSequence[] titles = {"Не уведомлять", "Первый раз", "Каждый раз", "Каждый день", "Каждую неделю"};
        final String[] values = {TopicApi.TRACK_TYPE_NONE, TopicApi.TRACK_TYPE_DELAYED,
                TopicApi.TRACK_TYPE_IMMEDIATE, TopicApi.TRACK_TYPE_DAILY, TopicApi.TRACK_TYPE_WEEKLY};
        final int[] selectedId = {1};
        new AlertDialogBuilder(context)
                .setTitle("Добавление в избранное/подписки")
                .setSingleChoiceItems(titles, 1, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        selectedId[0] = i;
                    }
                })

                .setPositiveButton("Добавить", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogInterface, int i) {
                        dialogInterface.dismiss();
                        String emailtype = values[selectedId[0]];

                        Toast.makeText(context, "Запрос на добавление отправлен", Toast.LENGTH_SHORT).show();
                        if (topicListItemTask != null) {
                            topicListItemTask.execute(emailtype);
                        } else {
                            final String finalEmailtype = emailtype;

                            new Thread(new Runnable() {
                                public void run() {

                                    Exception ex = null;

                                    String res = null;
                                    try {
                                        res = TopicApi.changeFavorite(Client.getInstance(), topicId, finalEmailtype);
                                    } catch (Exception e) {
                                        ex = e;
                                    }

                                    final Exception finalEx = ex;
                                    final String finalRes = res;
                                    handler.post(new Runnable() {
                                        public void run() {
                                            try {
                                                if (finalEx != null) {
                                                    Toast.makeText(context, "Ошибка добавления в избранное/подписки", Toast.LENGTH_SHORT).show();
                                                    Log.e(context, finalEx);
                                                } else {
                                                    Toast.makeText(context, finalRes, Toast.LENGTH_SHORT).show();
                                                }
                                            } catch (Exception ex) {
                                                Log.e(context, ex);
                                            }

                                        }
                                    });
                                }
                            }).start();
                        }
                    }
                })
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogInterface, int i) {
                        dialogInterface.dismiss();
                    }
                })
                .create().show();
    }


}

File: app/src/main/java/org/softeg/slartus/forpdaplus/listfragments/TopicUtils.kt
package org.softeg.slartus.forpdaplus.listfragments

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.text.TextUtils
import android.widget.Toast
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.MaterialDialog.ListCallback
import org.softeg.slartus.forpdaapi.FavTopic
import org.softeg.slartus.forpdaapi.Topic
import org.softeg.slartus.forpdaapi.TopicApi
import org.softeg.slartus.forpdaplus.App
import org.softeg.slartus.forpdaplus.Client
import org.softeg.slartus.forpdaplus.R
import org.softeg.slartus.forpdaplus.classes.ThemeOpenParams
import org.softeg.slartus.forpdaplus.classes.common.ArrayUtils
import org.softeg.slartus.forpdaplus.common.AppLog
import org.softeg.slartus.forpdaplus.db.TopicsHistoryTable
import org.softeg.slartus.hosthelper.HostHelper
import java.io.IOException

/**
 * Created by slinkin on 20.02.14.
 */
object TopicUtils {
    /**
     * С какими параметрами навигации юзер решил открывать топик:  view=getlastpost и тд
     */
    @JvmStatic
    fun getOpenTopicArgs(topicId: CharSequence?, template: CharSequence?): String? {
        val themeActionPref = getTopicNavigateAction(template)
        return getUrlArgs(topicId, themeActionPref, null)
    }

    /**
     * Какой тип навигации юзер выбрал:  getlastpost, getfirstpost и тд
     */
    private fun getTopicNavigateAction(template: CharSequence?): String? {
        return App.getInstance().preferences.getString(String.format("%s.navigate_action", template), null)
    }

    @JvmStatic
    fun getUrlArgs(topicId: CharSequence?, openParam: String?, defaultUrlParam: String?): String? {
        if (openParam == null) return defaultUrlParam
        if (openParam == ThemeOpenParams.BROWSER) return ""
        if (openParam == Topic.NAVIGATE_VIEW_FIRST_POST) return ""
        if (openParam == Topic.NAVIGATE_VIEW_LAST_POST) return "view=getlastpost"
        if (openParam == Topic.NAVIGATE_VIEW_NEW_POST) return "view=getnewpost"
        return if (openParam == Topic.NAVIGATE_VIEW_LAST_URL) {
            try {
                val historyTopicUrl = TopicsHistoryTable.getTopicHistoryUrl(topicId)
                if (TextUtils.isEmpty(historyTopicUrl)) "view=getlastpost" else Uri.parse(historyTopicUrl).query!!.replace("showtopic=\\d+&?".toRegex(), "")
            } catch (e: IOException) {
                e.printStackTrace()
                "view=getlastpost"
            }
        } else defaultUrlParam
    }

    fun getTopicUrl(topicId: String, urlArgs: String): String {
        return "https://${HostHelper.host}/forum/index.php?showtopic=" + topicId + if (TextUtils.isEmpty(urlArgs)) "" else "&$urlArgs"
    }

    @JvmStatic
    fun showSubscribeSelectTypeDialog(context: Context, handler: Handler,
                                      topic: Topic, action: ((String) -> Unit)? = null) {
        val titles = arrayOf<CharSequence>(context.getString(R.string.no_notify), context.getString(R.string.first_time), context.getString(R.string.every_time), context.getString(R.string.every_day), context.getString(R.string.every_week))
        val values = arrayOf(TopicApi.TRACK_TYPE_NONE, TopicApi.TRACK_TYPE_DELAYED,
                TopicApi.TRACK_TYPE_IMMEDIATE, TopicApi.TRACK_TYPE_DAILY, TopicApi.TRACK_TYPE_WEEKLY)
        var selectedSubscribe: String? = null
        if (topic is FavTopic) {
            selectedSubscribe = topic.trackType
        }
        val selectedId = intArrayOf(ArrayUtils.indexOf(selectedSubscribe, values))
        MaterialDialog.Builder(context)
                .title(R.string.add_to_favorites)
                .items(*titles)
                .itemsCallback(ListCallback { dialog, view, i, text ->
                    selectedId[0] = i
                    if (selectedId[0] == -1) return@ListCallback
                    val emailtype = values[selectedId[0]]
                    Toast.makeText(context, R.string.request_sent, Toast.LENGTH_SHORT).show()
                    if (action != null) {
                        action(emailtype)
                    } else {
                        Thread {
                            var ex: Exception? = null
                            var res: String? = null
                            try {
                                res = TopicApi.changeFavorite(Client.getInstance(), topic.id, emailtype)
                            } catch (e: Exception) {
                                ex = e
                            }
                            val finalEx = ex
                            val finalRes = res
                            handler.post {
                                try {
                                    if (finalEx != null) {
                                        Toast.makeText(context, R.string.error_request, Toast.LENGTH_SHORT).show()
                                        AppLog.e(context, finalEx)
                                    } else {
                                        Toast.makeText(context, finalRes, Toast.LENGTH_SHORT).show()
                                    }
                                } catch (ex: Exception) {
                                    AppLog.e(context, ex)
                                }
                            }
                        }.start()
                    }
                }) //.positiveText("Добавить")
                .negativeText(android.R.string.cancel)
                .show()
    }
}